src/DCON/Ode.jl

```diff
@@ -0,0 +1,722 @@
+    println(crit_out_unit, "step info")
+    if crit_break
+        write(crit_bin_unit)
+    end
+
+end
+
+# Example stub for kinetic crossing
+function ode_kin_cross()
+    # Implement kinetic crossing logic here
+    return
+end
+
+# Example stub for resistive crossing
+function ode_resist_cross()
+    # Implement resistive crossing logic here
+    return
+end
+
+function ode_step(ising::Int, odet::OdeState, equil::Equilibrium.PlasmaEquilibrium, intr::DconInternal, ctrl::DconControl)
+    # ODE integrator step
+    # Compute relative tolerances
+    singfac_local = typemax(Float64)
+    if ctrl.kin_flag
+ #       if ising == 1 && intr.kmsing >= 1
+ #           singfac_local = abs(odet.psifac - kinsing[ising].psifac) /
+ #                           (kinsing[ising].psifac - psilow)
+ #       elseif ising <= intr.kmsing
+ #           singfac_local = min(
+ #               abs(odet.psifac - kinsing[ising].psifac),
+ #               abs(odet.psifac - kinsing[ising-1].psifac)
+ #           ) / abs(kinsing[ising].psifac - kinsing[ising-1].psifac)
+ #       end
+    else
+        if ising <= intr.msing
+            singfac_local = abs(intr.sing[ising].m - ctrl.nn * odet.q)
+        end
+        if ising > 1
+            singfac_local = min(
+                singfac_local,
+                abs(intr.sing[ising-1].m - ctrl.nn * q)
+            )
+        end
+    end
+    tol = singfac_local < ctrl.crossover ? ctrl.tol_r : ctrl.tol_nr
+    rtol = tol
+
+    # Compute absolute tolerances
+    for ieq in 1:2, isol in 1:odet.msol
+        atol0 = maximum(abs.(u[:, isol, ieq])) * tol
+        if atol0 == 0
+            atol0 = typemax(Float64)
+        end
+        odet.atol[:, isol, ieq] .= ComplexF64(atol0, atol0)
+    end
+
+    # Choose psiout
+    if ctrl.node_flag
+        while odet.psifac < equil.sq.xs[ix]
+            ix += 1
+        end
+        psiout = equil.sq.xs[ix]
+        psiout = min(psiout, intr.psimax)
+        rwork[1] = psiout
+    else
+        psiout = intr.psimax
+    end
+
+    # Advance differential equations
+    odet.istep += 1
+
+    # Use DifferentialEquations.jl for general ODE solving in Julia
+    # 
+
+    # Define the ODE function in the DifferentialEquations.jl format
+    function ode_func!(du, u, p, t)
+        sing_der(neq, t, u, du)
+    end
+
+    # Set up the problem
+    u0 = copy(u)
+    tspan = (odet.psifac, psiout)
+    prob = ODEProblem(ode_func!, u0, tspan)
+
+    # Set tolerances
+    abstol = maximum(abs.(atol))
+    reltol = rtol
+
+    # Solve the ODE
+    sol = solve(prob, abstol=abstol, reltol=reltol)
+
+    # Update u and psifac with the solution at the end of the interval
+    u .= sol.u[end]
+    odet.psifac = sol.t[end]
+
+end
+
+function ode_unorm(intr::DconInternal, odet::OdeState, dout::DconOutput, fNames::DconFileNames, sing_flag::Bool)
+    # Compute norms of first solution vectors, abort if any are zero
+    odet.unorm[1:intr.mpert] .= sqrt.(sum(abs.(u[:, 1:intr.mpert, 1]).^2, dims=1)[:])
```

Comment:
The array inside sum will be allocated. You can do `sum(x -> abs(x) ^ 2, u[:, 1:intr.mpert, 1]), dims=1)` I think. It definitely works without the `dims=1`, but might take some testing to confirm.

Also not sure about that final `[:]`

Likewise below